### DISCUSSION TITLE
Correctly parsing the `-e` command line argument

### DISCUSSION THREAD

**Discussion by ofseed** [ghostty-org/ghostty#2633] [Feature Requests, Ideas] [answered]
ghostty currently cannot correctly parse (or say in the same way as other terminals) the command line argument `-e`, which leads to a common case where setting it as the default terminal via `xdg-terminal-exec` fails to open a file with a path that contains spaces or parentheses when using `nvim.desktop` as the default editor.

### Situation 1
Pass a parameter with spaces to `ghostty -e`, such as a path with spaces
```sh
ghostty -e nvim "path with spaces"
```
#### Expected Behavior
nvim opens a file at `"path with spaces"`.
#### Actual Behavior
Open the three files `"path"`, `"with"`, and `"spaces"` in nvim.

### Situation 2
Pass a parameter with parentheses to `ghostty -e`, for example
```sh
ghostty -e nvim "path with ()"
```
#### Expected Behavior
nvim opens the file `"path with ()"`
#### Actual Behavior
```
/bin/sh: -c: line 1: syntax error near unexpected token `('
/bin/sh: -c: line 1: `nvim path with ()'

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

Ghostty failed to launch the requested command:

/bin/sh -c nvim path with ()

Runtime: 15 ms
Exit Code: 2

Press any key to close the window.
```

**Comment by mitchellh** [comment]
Fully supportive of figuring out _some answer_ to this. I've converted to a planning discussion until we have an actionable result that we want to take. I also want to note there is some discussion already in a previous PR at #2620. 

I'll use this comment to provide additional background:

I've maintained multiple widely used projects in the past that do shell parsing, and it becomes a very hairy nightmare really quickly. My goal with Ghostty's `command` and `-e` features are that they have straightforward, well-defined cases they work and don't work. I don't have a strong opinion _where_ those edges are, except that maintainability is key. 

Second point of background: Ghostty originally _didn't support args at all_. You can probably use git blame to find the issue that brought this but originally we only allowed _single executables_ to completely bypass this problem. This was unliked for somewhat obvious reasons, but it had very clear semantics at least. 

My attempt to just use `/bin/sh -c` was to avoid shell parsing (hence passing the command through a shell to do the shell parsing). This breaks for the cases noted in the original post. 

I'm mostly curious how other terminals do it. Note we can't copy the source of a GPL licensed project so we have to be careful, but if anyone can conceptually explain how they work we can mimic that behavior if we find it maintainable.

**Reply by ManuelSchneid3r** [reply]
IIUC `xterm -e` compatibility does not involve shell parsing. Just take arguments after `-e` and use it as `argv` of the process to be run in the terminal.

**Reply by c-wicklein** [reply]
@mitchellh , is part of the problem that Ghostty has two distinct mechanisms for accepting an arbitrary command to run, the command line option `-e` and the configuration file option `command`?

I agree with @ManuelSchneid3r.  When an arbitrary program `A` run from the command line accepts among its arguments a directive to execute another program `B`, program `A` simply needs a sentinel argument indicating that the argument following it both identifies the program to execute and begins `B`'s argument vector.  There's nothing to parse, and I think even trying is a losing game.  Reasonable people who still need something to argue about can argue about what the sentinel argument should be.  (I find `-e` unintuitive because I think of switches as being commutative and any positional arguments which follow as having some prescribed order.  My general preference for programs which wrap the execution of other programs is for the first occurrence of `--` to indicate that any further arguments are for someone else.)

I think it's the second mechanism for specifying an arbitrary command to execute, the configuration file option `command` which creates confusion.  In this case, the command is no longer provided as a vector of null-terminated strings; it's just some text, perhaps the sequence of characters with non-inclusive boundaries of `=` and some line terminator.  Here, I think the _worst_ choice would be to reproduce the parsing behavior of a shell (which shell?).   Switching to a configuration format which can unambiguously represent a vector of strings would solve the problem, but I think there's a better way.  While it's not completely unreasonable for Ghostty to make some attempt at executing the given command directly (which is straightforward if the entirety of the command is `/usr/local/bin/my-hobby-shell`), in #3030 I've made a pitch for Ghostty to handle command execution by executing an instance of the user's login shell and writing the configured command to that shell before turning things over to the user for interaction.  I believe this is how Terminal.app works on macOS, and it has the side benefit (at least on macOS where this is a real problem) of ensuring that the arbitrary command Ghostty executes has a "fully configured environment" matching that of the user's login shell.

**Reply by ManuelSchneid3r** [reply]
I second the proposal to use `--` for the given reasons.

Generally I don't think there is much to discuss. This way to handle the argv to execute a process in a terminal is a defacto standard supported by all terminals I used so far. Furthermore there will (probably) be the xdg-terminal-execute standard that builds on top of this behavior.

**Reply by c-wicklein** [reply]
After some thought and work on #3030, I'm going to back pedal and suggest that in the balance, accepting the single argument following the `-e` switch as the command to run makes more sense than allowing the command to span multiple arguments, but my reasoning is also tied to #3030.

If Ghostty were to execute the alternative command with e.g. `execv`, as it does now, then what's proposed here would make things simpler, and the thorny problem would be how to specify a multi-argument command in the configuration file.  However, were Ghosty to do what's proposed in #3030 and treat the alternative command as a simple sequence of bytes (i.e a string) to send to the user's shell, then the configuration file is no longer a problem, and if Ghostty accepted a multi-argument alternative command on the command line, it would be necessary to join those arguments into a single string, respecting the quoting conventions of the user's shell.  Boo.  So, in that world, it makes far more sense for the command line argument `-e` to be followed by a single string which can be written to the user's shell, and the problem of quoting to create multi-argument commands is handed back to the user.

For a (contrived) example...

```
dir="/Users/me/Music and Movies"
ghostty -e "ls '-lart' '$dir'"
```

...passes Ghosty a single string, `ls '-lart' '/Users/me/Music and Movies'`, which Ghostty could write to the user's shell, without having to understand quoting.

**Reply by ManuelSchneid3r** [reply]
This effectively will not allow to run a command in ghostty without a shell. This is probably not what users would expect. and opens the door for tons of side effects. is this really necessary? if people want to have a shell in between they could still run  eg `ghostty -- /bin/sh -c htop`.

**Reply by c-wicklein** [reply]
@ManuelSchneid3r  A fundamental concern is that there are two methods of specifying an alternative command.  One (`ghostty`'s command line) is conducive to accepting a vector of arguments; the other (in the configuration file) is not.  Is it okay for `-e` and `command` to be different in this way?

**Reply by ManuelSchneid3r** [reply]
okay for what/whom?

**Reply by c-wicklein** [reply]
Having two different mechanisms for specifying an alternative command, a command line switch `-e` and the configuration file parameter `command`, where their semantics appear to differ (the former accepting multiple arguments and the latter accepting a single string) was bothering me.  Having stewed on it a bit, though, I think this is a small harm, if it's a harm at all, and it's more important that each mechanism work in an independently  sane way.  For the command line switch `-e`,  I'm back to thinking that the `xterm` model where `-e` must be last and a vector of arguments follows is appropriate.  For the configuration file parameter `command`, I still think executing a login instance of the user's login shell and writing the command to that shell rather than attempting to "execute" the command should be supported (ala `Terminal.app` on macOS.)

**Comment by SHU-red** [comment]
Hi there,
trying to use ghosty with shortcuts to open and execute a bash-script, taking a single argument but it does not seem to work

this works
```
~/Scripts/FuzzyExecute.sh ~/Scripts
```

this errors
```
ghostty -e "~/Scripts/FuzzyExecute.sh ~/Scripts"
```

Error in second terminal window
```
/bin/sh: line 1: /home/shured/Scripts/FuzzyExecute.sh: No such file or directory

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

Ghostty failed to launch the requested command:

/bin/sh -c ~/Scripts/FuzzyExecute.sh ~/Scripts

Runtime: 2 ms
Exit Code: 127

Press any key to close the window (although the file exists)
```

Not sure if its OK to add to this post of if the root cause belongs to a different issue/PR/discussion

**Comment by SHU-red** [comment]
I found a bit of a hacky solution for me (not being an expert)
Somehow by utilizing `fish -c` inbetween i got my script running but passing `~/Scripts` as argument did not work
So i changed my script using Env-Vars

So the main takeaway here is that i start a new fish shell via ghostty to run my bash script successfully
Below from my hyprland.conf because i need this as "quick start" script

```
bind = $mainMod, Y, exec, export FE_DIR=/home/shured/Scripts; $terminal -e fish -c "~/Scripts/FuzzyExecute/FuzzyExecute.sh"
```

Maybe not helpful for you but in case yes i wanted to write it down

Thanks

**Comment by m4r1vs** [comment]
`alacritty -e "echo hi` fails since `echo` is a shell built in and not an executable. It seems like other terms behave similarly. 
If I want the terminal to launch the command with a shell, I would expect the user to pass in the shell to the argument explicitly.

I would strongly welcome it if ghostty also did not launch everything through /bin/sh (or cmd.exe) but just implemented parsing of the given command properly. 

@mitchellh Maybe we could use something like [wordexp](https://man7.org/linux/man-pages/man3/wordexp.3.html) for POSIX and [CommandLineToArgvW](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-commandlinetoargvw) on windows?

**Reply by m4r1vs** [reply]
Note:
```
❯ gnome-terminal -e "ls"
# Option “-e” is deprecated and might be removed in a later version of gnome-terminal.
# Use “-- ” to terminate the options and put the command line to execute after it.
```

**Reply by m4r1vs** [reply]
```
❯ gnome-terminal -- "echo $SHELL"
# Error: Failed to execute child process “echo /usr/bin/zsh”: Failed to execve: No such file or directory
```

**Reply by m4r1vs** [reply]
However
```
❯ gnome-terminal -- echo $SHELL
❯
```

**Reply by m4r1vs** [reply]
Experimenting with wordexp [here](https://github.com/m4r1vs/ghostty/tree/wordexp-experiments) and currently running [this](https://github.com/m4r1vs/ghostty/tree/simple-shell-parser) branch which works for my needs.

**Reply by mitchellh** [reply]
The implementation of `wordexp` in libc on macOS execs out to perl, and in musl and gnu libc it execs out to `/bin/sh`. I'm not quite sure that's a nice improvement over what we're doing, although it does punt it to someone else's problem. 😆 I want to solve this so so I will continue to consider this.

**Comment by mitchellh** [comment]
Coming back to this, I'd really like to solve this. 

One suggestion above was to use `wordexp(3)`. I'm not a fan of this for reasons well stated in this blog post: https://www.cipht.net/2023/11/21/worried-by-wordexp.html tl;dr it is extremely slow and complex, all major implementations shell out anyways, and it enables arbitrary program execution during expansion.

The original issue as stated was that the `-e` command is not correctly parsed. `-e` comes from a shell or manual exec call which by definition already gives us a nice array of strings. I agree in this scenario there is no reason to run this through `/bin/sh`. 

Going further into implementation details, `-e` is implemented by being syntactic sugar over `initial-command` (and some other configs that are irrelevant for this issue and documented for that config). 

I think that running a command through `/bin/sh -c` by default for `command/initial-command` _in the config file_ results in the [principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment). But for `-e`, the opposite is true. 

## Proposal

What I'd like to propose is some syntax for `command/initial-command` to NOT run it through `/bin/sh -c`. The syntax I'm thinking would be to prefix the command with `!`. i.e. `command = ! /bin/whatever with args`.

This syntax isn't obvious and you'd definitely have to read documentation to understand what's going on, but my expectation is that _most users won't ever have to_ because both usage (in config file or `-e`) will apply the principle of least surprise. 

For the config file ones, we will NOT do any shell expansion. We accept the args space-separated as-is. That means the following would NOT work: `! nvim "file with spaces"` (you'd have to drop the `!`). But, since this is an opt-in, the user is shooting themselves in the foot. **To be clear, this would work with `-e` because it has a different default, so the original issue would be resolved.**

An alternative would be introducing a new config like `command-uses-shell` but I don't want to do this because we'd have to do it _twice_ (for command and initial command), which is really bleh to me.

**Reply by hqnna** [reply]
What about something similar to keybind actions? Where we have different "types" of `command` arguments, i.e. we have a `shell` type and a `direct` type (for lack of better wording). The `shell` type would be as things are now, where they are called by `/bin/sh`, while as the `direct` type would instead directly call the given command.

### Shell Commands
This would be the default and documentation and defaults would be updated to show this.
```ini
command = shell:ls -la
```
This would be like how things are as of current, so with our status quo.
```ini
command = ls -la
```

### Direct Commands
Unlike the former this tells ghostty to call the given command *directly* rather than passing it through `/bin/sh`.
```ini
command = direct:/bin/echo 'hello world'
```
This would instead not be passed into `/bin/sh`, and ghostty would instead directly call `/bin/echo` with the arguments.

---

Regarding passing arguments, in my experience programs that do this sort of thing, such as the helix text editor, dockerfiles, and similar take arguments as an *array* rather than as a string, this allows the user to split up the arguments in the config itself to make it clear what is seperate, i.e. in a dockerfile:
```dockerfile
CMD ["/bin/echo", "hello world"]
```
Keeping this in mind, and given we don't want to introduce much more syntax, we can treat spaces (that are not inside quotes, mind you) as separate arguments and form an array using this arguments to pass to the process.
```ini
command = direct:/bin/ls -la $HOME/documents
```
would be parsed into the following for example:
```
direct:
/bin/ls
-la
$HOME/documents
```
Of course the `direct` would be taken in ghostty itself to decide which type of command to use. Regarding env variables, we could get the current environment with zig's `EnvMap` type and deconstruct variables ourselves when not using the `shell` commands.

---

This would give a clear and concise way to tell the user which command type is being used as well as easy parsing for things.

**Reply by jcollie** [reply]
I think that this is the wrong approach. The problem really stems from the fact that when Ghostty encounters `-e` on the CLI it takes the array of strings that have _already_ been parsed by a shell, and then turns it _back into_ a single string:

https://github.com/ghostty-org/ghostty/blob/6f7977fef186faa9b9afe7707dc21a2eff59883b/src/config/Config.zig#L3151-L3155

What if we modify `initial-command` to be something more like this:

```
@"initial-command": ?union(enum) {
 parsed: [][]const u8,
 unparsed: []const u8
} = null;
```

If `initial-command` is encountered in a config file (or as `--initial-command="blah blah" on the CLI), then a single unparsed string would be set, and it would be run with `sh -c "<command>"`. However, if `initial-command` is set by `-e` (or by a DBus or other API call) then we'd store the pre-parsed strings and it would be run directly _without_ adding `sh -c`.

That way no fancy parsing of the config file is needed.

**Reply by mitchellh** [reply]
@jcollie This is exactly my proposal. I'm sorry if I worded it strangely but... that's exactly what I proposed. The `!` syntax (bikesheddable as Hanna did) is just to toggle that enum key. And `-e` would be modified to always set the "parsed" key (using your terminology).

**Reply by mitchellh** [reply]
@hqnna I think your proposal makes sense! It would definitely match other behaviors in our config much more closely. I also agree with the terminology you used, which coincidentally is the same terminology I used for my union tags in my WIP branch!

**Reply by jcollie** [reply]
I think that the `!` syntax (or anything else) is unnecessary unless we go all the way and support quoting and other escapes (or some other way of specifying arguments with spaces or other weirdness in them). Anything less I think is just going to generate bug reports.

I think one way that fits _a little_ more with the "Zen of Ghostty" would be to allow specifying `initial-command` multiple times to build up an array:

`initial-command = nvim`
`initial-command = file with spaces`

Commands with multiple elements in the array _would not_ use `sh -c` where single element arrays would.

**Reply by hqnna** [reply]
> I think that the `!` syntax (or anything else) is unnecessary unless we go all the way and support quoting and other escapes (or some other way of specifying arguments with spaces or other weirdness in them). Anything less I think is just going to generate bug reports.
> 
> I think one way that fits _a little_ more with the "Zen of Ghostty" would be to allow specifying `initial-command` multiple times to build up an array:
> 
> `initial-command = nvim` `initial-command = file with spaces`
> 
> Commands with multiple elements in the array _would not_ use `sh -c` where single element arrays would.

the issue with this approach is things that some things that supported arrays, i.e. `font-feature`, also now support parsing it as an array of arguments with one call, i.e. `font-feature = -liga -calt -dlig`. In my opinion using a syntax to *clearly* tell the user when it is or is not using the shell is a lot of a better UX than the user having to figure out that they need to specify multiple `initial-command` options.

**Reply by jcollie** [reply]
Yeah, Ghostty is a bit schizophrenic on how arrays are specified and doing it on multiple lines is kinda ugly (especially for this purpose). If we're going to allow it from the config file I think that we shouldn't half-ass it and fully support quoting (both single and double quotes) and backslash escapes.

Environment substitution would be nice as well but could result in some _interesting_ bug reports as  we would only be able to substitute environment variables that Ghostty knows about. That could be _very_ different from what the user expects depending on how Ghostty was launched.

Maybe limiting "substitution" to some small subset (like "%h" for home, "%d" for the working directory being inherited by the window) instead of a full environment substitution.

**Reply by mitchellh** [reply]
> I think that the ! syntax (or anything else) is unnecessary unless we go all the way and support quoting and other escapes (or some other way of specifying arguments with spaces or other weirdness in them). Anything less I think is just going to generate bug reports.

I disagree with all three points here:

1. **`!` being unnecessary.** Our CLI arguments have always mapped to file-based config. Everything you can do in the CLI you can do in the config and vice versa. The syntax is mostly there to support this as a non-magical behavior for `-e`. `-e` is the only real "exception" to the direct isomorphism but it still cleanly and deterministically maps to something you could do manually via a file.

2. **"unless we go all the way and support ... escapes".** This isn't necessary because `!` is explicitly documented as NOT supported escapes/expansion. You'd just avoid the `!` (the default) to get all that by running through a shell as we do today. Nothing changes for users.

3. **This will generate bug reports.** Its an opt-in behavior. Nothing changes _by default_ for anyone except `-e` users (which you already support). If a user is explicitly using `!` which is documented as "not supporting shell expansion" and also documented that you simply remove the `!` to get shell expansion, well, that's an easy close. I doubt we'll see any increase in bugs here.

**Reply by mitchellh** [reply]
> Maybe limiting "substitution" to some small subset (like "%h" for home, "%d" for the working directory being inherited by the window) instead of a full environment substitution.

No, I'm very against any form of this. My original post explains the motivation:

> My goal with Ghostty's command and -e features are that they have straightforward, well-defined cases they work and don't work. I don't have a strong opinion where those edges are, except that maintainability is key.

Namely, I'd rather have a very clear boundary of what is and isn't supported than a vague scale. 

I've implemented shell parsing at various points on the scale in a half dozen well adopted tools over the past 15 years and it was never once worth it. You always get users that need that "one more thing" ("it's only home expansion!") or you get a flurry of bug reports that things don't work like their shell does because its a subset. For most of those projects after a decade we reverted everything and went back to "dumb space splitting" which worked _wonderful_. The bugs disappeared. That's why I'm suggesting `!` does dumb string splitting. I've done it before. It works fantastically.

Going all the way and subprocessing to `/bin/sh` or doing nothing at all are the only sane options. This is even bolstered by libc itself which subprocesses to `/bin/sh` to do expansion! (The only libc API that does so, I believe!)

EDIT: I think Gnome reimplements [a subset] of shell parsing for their "glib" library, and the implementation is almost 3000 lines of C code. And if you put a command expansion in there, it just subprocesses to `/bin/sh` anyways. I haven't looked at that in awhile, I only know about it because like 10 years ago I used it as a basis for my own implementation (that I ultimately threw away).

**Reply by c-wicklein** [reply]
I like where `-e` seems to be headed, but I don't follow the distinction between `command = ...` and `command = ! ...`.  With or without the `!`, I think the proposal still involves Ghostty executing what is specified with `command`.

I like the idea of using `!` to signal an alternative mechanism for handling `command` and `initial-command`.  This essentially matches how `Terminal.app` is configured on macOS. A command to run at start up can be specified, and its default mechanism of execution is changed by selecting a checkbox.

![image](https://github.com/user-attachments/assets/8b16cecc-9613-4667-a04c-5aa06a56c670)

In patches relative to Ghostty 1.1.2, I've introduced alternatives to `command` and `initial-command` called `send-command` and `initial-send-command`, but I don't like this solution as much as using `!`.  I was addressing my own problem without really understanding Ghostty's architecture or having any prior exposure to Zig.  In my build, setting the configuration option `send-command` is essentially the same thing as setting the option `command` and checking the "Run inside shell" box.

Other than patching `Config.zig` to include the options `send-command` and `initial-send-command`, the only other change I make is to handle them in `Surface.zig`.

```
--- src/Surface.zig.orig        2025-02-13 17:03:54
+++ src/Surface.zig     2025-02-26 07:50:47
@@ -687,8 +687,30 @@
         }
     }

+    const send_command: ?[]const u8 = if (app.first)
+        config.@"initial-send-command" orelse config.@"send-command"
+    else
+        config.@"send-command";
+
     // We are no longer the first surface
     app.first = false;
+
+    const cmd = send_command orelse return;
+    if (cmd.len == 0) {
+        return;
+    }
+
+    var buf = try self.alloc.alloc(u8, cmd.len + 1);
+    defer self.alloc.free(buf);
+    for (cmd, 0..) |ch, i| {
+        buf[i] = ch;
+    }
+    buf[buf.len-1] = '\r';
+
+    self.io.queueMessage(try termio.Message.writeReq(
+        self.alloc,
+        buf,
+    ), .unlocked);
 }

 pub fn deinit(self: *Surface) void {
```

What I'd like to see from Ghostty so that I won't have to perpetuate my patches is:

1. `-e` works as with `xterm`.  I think @mitchellh is leaning that way. 
2. By default , `command` / `initial-command` continue to involve some variation on taking arguments to e.g. `execv` from the specified string.  I'm not too concerned with exactly how that works.
3. Optionally, if the value assigned to `command` /  `initial-command` begins with `!`, then that leading `!` is stripped and the remainer of the string is "executed" by *writing* it to a login instance of the user's login shell, matching in effect the behavior my builds have with `send-command` / `initial-send-command`.

For me, writing commands to a login shell is a coup.  The terminal program needs no special knowledge of how any particular shell parses commands, the specified command is parsed by the shell running it, i.e. correctly, and arguably, commands are executed in the least surprising way possible -- opening a new terminal with `command` set to `some command` has *exactly* the same effect as opening a new terminal window without having set `command` and me then typing `some command` into the terminal and pressing return.

**Reply by mitchellh** [reply, marked as answer]
Promoted to issue #7032
